### ISSUE TITLE
Release automation

### ISSUE THREAD

**Issue by jeanas** [pygments/pygments#2585]
Things I'd like to automate:

* From tag to PyPi -- ideally, to test-pypi on every tagged commit (https://github.com/marketplace/actions/pypi-publish) -- and the actual release would be a special action I just click on. It's not that it takes a lot of time, but I'm always nervous I mess something up with the command line, forget to delete a file, git clean, etc. -- I very diligently work through the `release-checklist` to avoid that. Literally signing things off.
* Auto-formatting -- I tend to clean up the formatting of the lexers every time close to release, at least the worst offenders. I use `autopep8` at the moment, would rather apply `flake8` or `black` on the entire codebase.
* Auto-check that the new arguments like URL etc. are present on new lexers
* Auto-check `.. versionadded:: ` is there -- costs me _a lot of time_ to open up every Lexer close to release and make sure it's present and in the right format (i.e. `2.17.0` vs. `2.17`)
* Actually get all checks working/passing (i.e. the additional checkers I wrote and possibly PyLint). `check_whitespace_tokens` and `check_repeated_tokens` need an expected-fail list so we can whitelist currently existing lexers until we fix those, but new lexers should always pass those tests.
* Verify all PR numbers closed/merged since last release are mentioned in the `CHANGES` file. I'm pretty good at assigning tasks to milestones now, but I still miss things in the `CHANGES` file, and it's super time consuming to open 100 tabs, go through each item one-by-item, check the PR number/issue number is present, etc. If there was a way to auto-generate the changelog that would be even better, but my experience is that those look pretty ugly and some manual checkup is fine.

I'll get to the release in a moment, thanks for the offer though!

_Originally posted by @Anteru in https://github.com/pygments/pygments/pull/2576#discussion_r1398348679_

**Comment by Anteru** [comment]
Yes. The problem is that we haven't been doing this consistently though, so I'd like _one_ file which contains all exclusions. I.e. run this one over all lexers, find which checkers work/don't work, then write it out with all exclusions, so we can run it for new ones while reducing the exclusion list. That requires however unifying all the checkers somehow or port them to the test framework, and I haven't had enough time to figure out how to integrate all of this with `pytest`. I'm sure it's doable.

**Comment by birkenfeld** [comment]
Yes, Ruff is really good but its linting only completely replaces pyflakes for now. pylint catches way more things (also way more false positives, sigh). A clean Ruff run could be part of CI.

For formatting, if we ever wanted to do that, Ruff is also a good option because it allows selecting single quotes as the preferred style :smile:

**Comment by Anteru** [comment]
Ok, I also prefer single quotes for strings, so I guess Ruff it is. I'll give it a try.

**Comment by Anteru** [comment]
I don't have a strong opinion either. I'm not a huge fan of Black's style, but realistically, what are our options? I'm extensively using `flake8` when editing Python code (and try to fix warnings it finds in Pygments as well), and without some automated code formatting, I'll spend the rest of my life telling people about styling issues. Like, I'd really like to enforce 79-wide lines so I can diff two things side-by-side and making this automatic is much less work than arguing with every single committer about the width of their screen and how much more productive they get when writing 95 character long lines.

I've been doing the same at work, you force `clang-format` on everything, and it just saves time in the end. It also makes other linters etc. more useful as my VS Code will only show "real" issues afterwards, and not mix in the formatting issues.

**Comment by birkenfeld** [comment]
> Lol. I don't have marked opinions on formatting, but I thought you didn't want Black due to its style? (Ruff's formatter is designed to produce output almost identical to Black.)

Yeah I didn't want to suggest to implement it immediately, and possibly there will never be a formatter that keeps the lexer code compact enough for my tastes...

**Comment by Anteru** [comment]
In my opinion we're just delaying the inevitable here. Will it be ugly? Probably. Does it mean we have a few cases of `foo|bar|baz` which are better off using `words(...)`? For sure. But what's the alternative -- delay auto-formatting indefinitely, and end up doing some small fixes every once in a while?

**Comment by birkenfeld** [comment]
I also despise this waste of space:
```
(
    regex,
    token
),
(
    regex,
    token
)
```

**Comment by hugovk** [comment]
Black would reformat that to:

```python
(regex, token),
(regex, token)
```

It would only expand to one argument per line if they don't all fit in the line limit.

Or, if you manually added the "magic trailing comma" after `token` and it'll assume you actually want that and expand it: 

```python
(
    regex,
    token,
),
(
    regex,
    token,
)
```
But don't add those if you don't want that :)

**Comment by EpicWink** [comment]
> From tag to PyPi -- ideally, to test-pypi on every tagged commit (https://github.com/marketplace/actions/pypi-publish)

+1 on using the `pypi-publish` action, using Trusted Publishing (with attestations). This means I can know for certain that what I download from PyPI is the same artefact which was generated in GitHub CI, and that what I see in GitHub is the same as what is installed - handy for auditing (rather than having to manually review all of the installed files on each release).

To do this, you would add a "PyPI publish" job to the `build` CI workflow (perhaps conditioned on Git tag, etc), which simply downloads the wheel and sdist artefacts, then uploads using the pypi-publish GitHub action.

See [the Python packaging documentation](https://packaging.python.org/en/latest/guides/publishing-package-distribution-releases-using-github-actions-ci-cd-workflows/#configuring-trusted-publishing), [the PyPI documentation](https://docs.pypi.org/trusted-publishers/), and [the official pypi-publish GitHub action documentation](https://github.com/pypa/gh-action-pypi-publish?tab=readme-ov-file#trusted-publishing) on trusted publishing - you'll need to configure an environment in PyPI and GitHub.

**Comment by lmmx** [comment]
In the wake of the (ongoing) PyPI supply chain attacks on litellm/telnyx repos (in just the last week) I would also like to express support for Trusted Publishing. pygments is the 24th biggest package on PyPI by downloads, and ~half of the top 100 packages have adopted TP now, so there are lots of examples to help follow.

The litellm hack involved a GitHub Action being compromised which then was used to steal credentials including PyPI API keys being used for publishing on CI, which was then used (off of GitHub) to publish malicious versions of the litellm package. Then 2 days later a similar one with telnyx.

When you use Trusted Publishing there is no PyPI auth credential available for anyone to steal. Instead, GitHub vouches for the identity, and you pre-declare a particular CI workflow (by filename). It also requires creating an 'environment' for the repo on GitHub (suggested name "pypi") which you also register the package with on PyPI.

<sup>*</sup> Post-mortems of the litellm attack: 
1. https://futuresearch.ai/blog/litellm-pypi-supply-chain-attack/
2. https://snyk.io/articles/poisoned-security-scanner-backdooring-litellm/

**Comment by birkenfeld** [comment]
For me, this whole debacle makes me think that simply *not* releasing from CI/GitHub actions is not the worst solution after all. 

In any case, I don't support the ongoing integration of GitHub as an indispensable part of the pipeline.

**Comment by EpicWink** [comment]
> For me, this whole debacle makes me think that simply not releasing from CI/GitHub actions is not the worst solution after all.
>
> In any case, I don't support the ongoing integration of GitHub as an indispensable part of the pipeline.

@birkenfeld Having long-lived API tokens which are accessible to your build dependencies is worse IMO. Additionally, running `twine upload` locally won't signal to users of your packages that the publication was performed by a trusted entity.

There are other CI services supported by PyPI for trusted publishing (ActiveState, GitLab, Google Cloud; and [Codeberg (Forgejo)](https://discuss.python.org/t/new-oidc-providers-for-trusted-publishing/106334) is in discussion).

**Comment by lmmx** [comment]
The lesson I took from it was the need to do CI releases properly, not abandon them. The remediations were all well known good practices: 

1. SHA pinning your GH Actions rather than using tag names
2. Trusted Publishing
3. using dependency cooldowns in production environments (i.e. not upgrading package/Action versions without a fixed delay). In these cases even 1 day would have sufficed as PyPI quarantined the packages in 3h for litellm and 6h for telnyx (3d or 7d is typical).

My read was they had not taken basic security hardening measures in CI (which you can audit using [zizmor](https://zizmor.sh) FYI).

Trusted Publishing has been an alternative on PyPI since 2023 and would've meant no secret would be in CI to steal, which is ultimately what matters for package publishing. Lack of this was used by bad actors to publish packages in both litellm/telnyx hacks. Short-lived OIDC tokens scoped to a single release workflow, managed by the platform internally, and with more eyes on the source, seems more secure to me.

Using someone's laptop instead of CI is a less auditable point of failure, and incentivises bad actors to do social engineering. We are not seeing GitHub account takeover being used to publish packages via Trusted Publishing.

**Comment by Anteru** [comment]
Given how well GitHub is running, I would also not invest more of our energy on integrating GitHub even deeper. It's all a question of who you trust more: Me or Georg to safeguard our access token, or our GitHub account credentials, or GitHub's pipeline? There's no magic here that makes GitHub safe. You can always attest yourself if you want to that the package we published matches the tag by unpacking the wheel and comparing the files. We don't have binaries here, so every build is 100% deterministic and repeatable.

Social engineering is quite hard if there are only two-three people with publishing permissions who are not sitting in the same office. And the attack vector is the same: If someone successfully social engineers me to put in some code and publish to PyPI, the same person can social engineer me to make a commit on GitHub and press publish. What attack vector are we safeguarding against here? It seems like the only attack vector is the fact that I have a PyPI token on my "laptop" which I use to publish that I change with some frequency, and we're trading that for my GitHub account credentials which I also change with some frequency. The thing is, you're trying to protect against me or another maintainer doing something malicious, but that doesn't work because we already have the keys to the kingdom and can always do something malicious. If someone steals our access token, presumably they can also steal our email account password or GitHub account, too, and then do malicious things with that.

Also notice the irony that the PyPI credentials were leaked _because of CI_, quoting here:

> they obtained the maintainer's PyPI credentials through a prior compromise of Trivy, an open source security scanner used in LiteLLM's CI/CD pipeline.

**Comment by lmmx** [comment]
> the attack vector is the same

Trusted Publishing fundamentally averts the attack vector of credential theft -> package publishing hijack by removing stored PyPI API keys from the equation entirely. If the maintainer’s machine houses the auth credentials for package publishing it becomes the target for such attacks.

litellm got 120k installs before PyPI quarantined it ([source](https://ramimac.me/teampcp/)), if package maintainers’ were among those and ran an “infostealer” (here a malicious .pth file that auto-runs code in Python without importing anything, that tried to hoover up any credentials it could find), then PyPI API credentials would have produced a worm like the [Shai Hulud](https://blog.pypi.org/posts/2025-11-26-pypi-and-shai-hulud/) attack (which did not hit PyPI yet but their blog warns to prepare for by moving to Trusted Publishing for this reason). It also suggests requiring human approval before publishing (so you could harden this by having multiple people approve the release). To me that sounds more secure.

I think it is quite a feasible risk, all the more for devs installing software from NPM.

>  I have a PyPI token on my "laptop" which I use to publish that I change with some frequency

Credential rotation helps avert incidents where secrets are harvested in advance before use, yes, but would not avert a scenario where packages were quickly republished after credential breach (like litellm was).

> Notice the irony that the PyPI credentials were leaked because of CI

The trivy-action that got breached was installed in the litellm CI, but the hackers then siphoned credentials from the victims who installed the litellm package. Those victims could be in any Python environment, local or CI.

> You can always attest yourself

When I say attestation here I am referring to [PEP 740](https://peps.python.org/pep-0740/), the process that PyPA have in place for software developers to ‘sign’ the software cryptographically, not just reproducibility.

> The thing is, you're trying to protect against me or another maintainer doing something malicious

I wasn’t suggesting you or other maintainers going rogue so much as being compromised in one of these waves, and the PyPI API key being taken allowing a bad actor to publish a malicious version of the package on your behalf.

On that note, PyPI have just put out a security advisory also suggesting Trusted Publishing in the wake of litellm https://blog.pypi.org/posts/2026-04-02-incident-report-litellm-telnyx-supply-chain-attack/#trusted-publishers-over-api-tokens

> ### Trusted Publishers over API tokens
>
> If the platform you use to publish to PyPI supports [Trusted Publishers](https://docs.pypi.org/trusted-publishers/) (GitHub, GitLab, Google Cloud Build, ActiveState) then you should use Trusted Publishers instead of API tokens.
>
> PyPI API tokens are "long-lived", meaning if they are exfiltrated by an attacker that attacker can use the token at a much later date even if you don't detect the initial compromise. Trusted Publishers comparatively uses short-lived tokens, meaning they need to be used immediately and don't require manual "rotating" in the case of compromise.
>
> Trusted Publishers also provides a valuable signal to downstream users through [Digital Attestations](https://docs.pypi.org/attestations/). This means users can detect when a release hasn't been published using the typical release workflow, likely drawing more scrutiny.

**Comment by tschm** [comment]
We have developed rhiza-tools for this. I show you how "make release" looks like

```
thomasschmelzer@Thomass-MacBook-Pro rhiza % make release
# Ensure the /Users/thomasschmelzer/projects/jqr/repos/rhiza/bin folder exists
# Install uv/uvx only if they are not already present in PATH or in the install dir
/Users/thomasschmelzer/.local/bin/uvx "rhiza-tools>=0.3.3" release ;
Current branch: main
**? Would you like to bump the version before releasing? Yes**
**? Select bump type (Current: 0.8.21) Minor (0.8.21 -> 0.9.0)**
Checking remote status...
Preflight: tag 'v0.9.0' is available on remote
Bumping version to: 0.9.0
Using language: python
Current branch: main
Current version: 0.8.21
New version will be: 0.9.0

Files to be modified:
  Changes in pyproject.toml:
    Line 3:
      - version = "0.8.21"
      + version = "0.9.0"

Running preflight validation (dry-run)...
Preflight validation passed
Running bump-my-version...

✓ Version bumped: 0.8.21 -> 0.9.0

Modified files:
  • pyproject.toml

Don't forget to run 'uv lock' to update the lockfile if needed.
Current version: 0.9.0
Expected tag: v0.9.0
Tag 'v0.9.0' found locally
  Commit: tag v0.9
  Date: 2026-04-03 15:14:47 +0400
  Message: Chore: bump version 0.8.21 → 0.9.0
Preparing to push tag to remote...
Pushing tag 'v0.9.0' to origin will trigger the release workflow.

Commits included in this release (since v0.8.21):
  • 93b9e88 Chore: bump version 0.8.21 → 0.9.0
  • 59d5203 Update rhiza_sync.yml
  • aa76362 chore: merge origin/main into main, keep workflow consolidation from release/v0.8.21
  • ed8ddd1 chore(deps): update pre-commit hook jebel-quant/rhiza-hooks to v0.3.2 (#964)
  • 33051b2 chore(deps): update pre-commit hook astral-sh/ruff-pre-commit to v0.15.9 (#962)
  • 910c5c2 chore(deps): update dependency astral-sh/uv to v0.11.3 (#961)
**Push tag to remote and trigger release workflow? [y/N]: y**
Pushing bump commit to remote...

Pushing tag to remote...
Command: git push origin refs/tags/v0.9.0

✓ Release tag v0.9.0 pushed to remote!
The release workflow will now be triggered automatically.
Monitor progress at: https://github.com/Jebel-Quant/rhiza/actions
Release process completed successfully!
```

https://github.com/Jebel-Quant/rhiza-tools (this project is the cli for the rhiza project)
Every release is properly signed, see https://github.com/Jebel-Quant/rhiza-tools/releases/tag/v0.4.3

We do not support tokens :-)

**Comment by Anteru** [comment]
Note that we _don't_ use our platform to publish. The advice from PyPI is "if the platform you use to publish", but we actually publish separately, not from GitHub, and we want to avoid locking ourselves deeper into GitHub at this stage.

Also:

> Credential rotation helps avert incidents where secrets are harvested in advance before use, yes, but would not avert a scenario where packages were quickly republished after credential breach (like litellm was).

Note that both credentials are on the same machine. If you can breach my machine to obtain my PyPI token, you also have access to my PyPI login and my GitHub. We're toast in that case. This works nicely if you have multiple people who have different access levels, but for projects with a single or two maintainers, it won't help much (especially not if the approver, committer, and publisher are all the same person - it's the classic murderer, detective and victim in one case.)
